src/s2e/Plugins/ExecutionMonitors/FunctionMonitor2.cpp

```diff
@@ -30,38 +31,30 @@ void FunctionMonitor2::initialize() {
 
 void FunctionMonitor2::onTranslateBlockEnd(ExecutionSignal *signal, S2EExecutionState *state, TranslationBlock *tb,
                                            uint64_t pc, bool isStatic, uint64_t staticTarget) {
-    enum ETranslationBlockType tb_type = tb->se_tb_type;
-    if (!(tb_type == TB_CALL || tb_type == TB_CALL_IND)) {
+    if (!m_processDetector->isTracked(state)) {
         return;
     }
 
-    if (m_monitor->isKernelAddress(pc)) {
-        return;
+    if (tb->se_tb_type == TB_CALL || tb->se_tb_type == TB_CALL_IND) {
+        signal->connect(sigc::mem_fun(*this, &FunctionMonitor2::onFunctionCall));
     }
-
-    signal->connect(sigc::mem_fun(*this, &FunctionMonitor2::onExecuteStart));
 }
 
-void FunctionMonitor2::onExecuteStart(S2EExecutionState *state, uint64_t pc) {
-    if (!m_processDetector->isTracked(state)) {
```

Comment:
You can't remove this. The OS may share physical code pages between processes, so you must figure it out at runtime (and instrument every block as a consequence).

Reply:
It hasn't been removed, just moved to the translate callback (https://github.com/S2E/libs2eplugins/blob/ab338174eac24474d016ea1d6a00008916b380ec/src/s2e/Plugins/ExecutionMonitors/FunctionMonitor2.cpp#L34). The other plugins seemed to do this check earlier (at translation, rather than execution), so I did the same.

Reply:
You cannot move it to translate for the reason I mentioned. The other plugins are wrong if they do that (we might not have fixed them all).

Reply:
If you move it to translate, and the OS shares the same code between two processes, your instrumentation will end up being called in unrelated processes.

Reply:
The reason is this: https://github.com/S2E/libcpu/blob/master/src/cpu-exec.c#L90
TBs are indexed by pc + phys addr.

Reply:
Oh ok, cool! I thought today had not been very productive, but I've learnt something now! I'll fix this tomorrow (and have another look at the other plugins).

Thanks!